File: packages/han/lib/data/hosted/schema/hook-executions.ts
/**
 * Hook Executions Schema
 *
 * Tracks hook execution history for sessions.
 */

import {
  boolean,
  integer,
  pgTable,
  text,
  timestamp,
  uuid,
  varchar,
} from 'drizzle-orm/pg-core';
import { organizations } from './organizations.ts';
import { sessions } from './sessions.ts';

/**
 * Hook Executions table
 *
 * Records each hook execution with its outcome.
 */
export const hookExecutions = pgTable('hook_executions', {
  id: uuid('id').primaryKey().defaultRandom(),
  organizationId: uuid('organization_id')
    .notNull()
    .references(() => organizations.id, { onDelete: 'cascade' }),
  sessionId: uuid('session_id').references(() => sessions.id, {
    onDelete: 'cascade',
  }),
  orchestrationId: uuid('orchestration_id'),
  taskId: varchar('task_id', { length: 255 }),
  hookType: varchar('hook_type', { length: 100 }).notNull(), // 'Stop', 'SessionStart', etc.
  hookName: varchar('hook_name', { length: 255 }).notNull(),
  hookSource: varchar('hook_source', { length: 255 }),
  directory: text('directory'),
  durationMs: integer('duration_ms').notNull().default(0),
  exitCode: integer('exit_code').notNull().default(0),
  passed: boolean('passed').notNull().default(false),
  output: text('output'),
  error: text('error'),
  ifChanged: text('if_changed'), // JSON array of glob patterns
  command: text('command'),
  status: varchar('status', { length: 50 }).default('completed'), // 'pending', 'running', 'completed', 'failed'
  consecutiveFailures: integer('consecutive_failures').default(0),
  maxAttempts: integer('max_attempts').default(3),
  pid: integer('pid'),
  pluginRoot: text('plugin_root'),
  executedAt: timestamp('executed_at', { withTimezone: true }).defaultNow(),
});

/**
 * TypeScript type for a hook execution row
 */
export type HookExecution = typeof hookExecutions.$inferSelect;
export type NewHookExecution = typeof hookExecutions.$inferInsert;

File: packages/han-rs/crates/han-db/src/entities/hook_executions.rs
//! Entity: hook_executions (track every hook run for metrics)

use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "hook_executions")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub id: String,
    pub orchestration_id: Option<String>,
    pub session_id: Option<String>,
    pub task_id: Option<String>,
    pub hook_type: String,
    pub hook_name: String,
    pub hook_source: Option<String>,
    pub directory: Option<String>,
    pub duration_ms: i32,
    pub exit_code: i32,
    pub passed: i32,
    #[sea_orm(column_type = "Text", nullable)]
    pub output: Option<String>,
    #[sea_orm(column_type = "Text", nullable)]
    pub error: Option<String>,
    #[sea_orm(column_type = "Text", nullable)]
    pub if_changed: Option<String>,
    #[sea_orm(column_type = "Text", nullable)]
    pub command: Option<String>,
    pub executed_at: String,
    pub status: Option<String>,
    pub consecutive_failures: Option<i32>,
    pub max_attempts: Option<i32>,
    pub pid: Option<i32>,
    pub plugin_root: Option<String>,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::orchestrations::Entity",
        from = "Column::OrchestrationId",
        to = "super::orchestrations::Column::Id"
    )]
    Orchestration,
    #[sea_orm(
        belongs_to = "super::sessions::Entity",
        from = "Column::SessionId",
        to = "super::sessions::Column::Id"
    )]
    Session,
}

impl Related<super::orchestrations::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Orchestration.def()
    }
}

impl Related<super::sessions::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Session.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}
